### PR TITLE
[Dataset] Chameleon

### PR DESCRIPTION
**Pull request by mufeili**
## Description
This PR adds the variant of Chameleon as introduced in [Geom-GCN: Geometric Graph Convolutional Networks](https://arxiv.org/abs/2002.05287).

A few thoughts and suggested practice:
- Handling changes of the data source. In general, this should be very rare. If the data source indeed changes, the url will need to be changed as well. This PR appends a hash value based on URL to the data directory so that the data can be re-downloaded and processed by checking if the directory exists.
- DGLGraph construction
    - If the graph is relatively small, then constructing it from the raw data files is very efficient as in this PR. In this case, I suggest not caching any DGL-specific data like DGLGraph. This can save a lot of trouble like if the graph is constructed in the desired way.
    - If the graph is very large, then we should have some file to track the configuration associated as I suggested in #4293.
- Test and regression
    - For small graphs as in this PR, unit tests are cheap enough.
    - For larger graphs, then we can have very preliminary unit tests or even no unit tests at all and instead rely on regression tests.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/data/wiki_network.py
Comment (jermainewang) [inline python/dgl/data/wiki_network.py:1]: Suggest to avoid using framework-agnostic backend for new datasets. Make them PyTorch only.

File: python/dgl/data/wiki_network.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            test_masks += [F.zerocopy_from_numpy(f["test_mask"])]
+        self._g.ndata["train_mask"] = F.astype(
+            F.stack(train_masks, dim=1), F.data_type_dict["bool"]
+        )
+        self._g.ndata["val_mask"] = F.astype(
+            F.stack(val_masks, dim=1), F.data_type_dict["bool"]
+        )
+        self._g.ndata["test_mask"] = F.astype(
+            F.stack(test_masks, dim=1), F.data_type_dict["bool"]
+        )
```
Comment (jermainewang) [inline python/dgl/data/wiki_network.py:89]: The general question is whether the logic of data processing / graph construction should belong to DGL core codebase or not.

File: python/dgl/data/dgl_dataset.py [hunk 2]
```diff
@@ -221,6 +220,15 @@ def _get_hash(self):
...
         return hash_func.hexdigest()[:8]
 
+    def _get_hash_url(self):
+        """Compute the hash of the url"""
+        if self._url is None:
+            return ""
+        else:
+            hash_func = hashlib.sha1()
+            hash_func.update(str(self._url).encode("utf-8"))
+            return "_" + hash_func.hexdigest()[:8]
```
Comment (frozenbugs) [inline python/dgl/data/dgl_dataset.py:230]: the `"_" +` should not be appended here since the function is named. _get_hash_url

File: python/dgl/data/wiki_network.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
+"""
+Wikipedia page-page networks on three topics: chameleons, crocodiles and
+squirrels
```
Comment (frozenbugs) [inline python/dgl/data/wiki_network.py:3]: end with dot.

Comment (frozenbugs) [inline python/dgl/data/wiki_network.py:3]: I don't see crocodiles and squirrels related code.

File: python/dgl/data/wiki_network.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            raw_dir=raw_dir,
+            hash_key=(url),
+            force_reload=force_reload,
+            verbose=verbose,
+            transform=transform,
+        )
+
+    def process(self):
+        """Load and process the data."""
+        # node features and labels
```
Comment (frozenbugs) [inline python/dgl/data/wiki_network.py:50]: \# Process node features and labels.

File: python/dgl/data/wiki_network.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            data = [[int(v) for v in r.split("\t")] for r in data]
+        COO = F.swapaxes(F.tensor(data, dtype=F.data_type_dict["int64"]), 0, 1)
+        dst = F.squeeze(F.narrow_row(COO, start=0, stop=1), dim=0)
+        src = F.squeeze(F.narrow_row(COO, start=1, stop=2), dim=0)
+
+        self._g = graph((src, dst), num_nodes=F.shape(features)[0])
+        self._g.ndata["feat"] = features
+        self._g.ndata["label"] = labels
+
+        # split
```
Comment (frozenbugs) [inline python/dgl/data/wiki_network.py:73]: \# Split blablabla.

File: python/dgl/data/wiki_network.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+    @property
+    def num_classes(self):
+        return self._num_classes
+
+
+class ChameleonDataset(WikiNetworkDataset):
+    """Wikipedia page-page network on chameleons from `Multi-scale Attributed
+    Node Embedding <https://arxiv.org/abs/1909.13021>`__ and later processed by
+    `Geom-GCN: Geometric Graph Convolutional Networks
```
Comment (frozenbugs) [inline python/dgl/data/wiki_network.py:115]: Why Geom-GCN is related? Is this dataset designated for Geom-GCN?

File: python/dgl/data/wiki_network.py
Comment (mufeili) [reply python/dgl/data/wiki_network.py:1]: Done

File: python/dgl/data/wiki_network.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/data/wiki_network.py]: Upon the offline discussion, the general principle will be including the logic of data processing in the core DGL codebase.

File: python/dgl/data/dgl_dataset.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/data/dgl_dataset.py:230]: renamed to `_get_hash_url_suffix`

File: python/dgl/data/wiki_network.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/data/wiki_network.py]: removed for now, will add later in another PR after we merge this one

File: python/dgl/data/wiki_network.py (hunk 6 above)
Comment (mufeili) [reply python/dgl/data/wiki_network.py:111]: Geom-GCN introduced this variant of the dataset, including turning the task from node regression into node classification, modifying node features, and introducing these dataset splits

Comment (frozenbugs) [reply python/dgl/data/wiki_network.py:111]: Maybe reword it. Currently working sounds like this dataset is just used by Geom-GCN.
e.g. later processed by --> introduced by.
```
Wikipedia page-page network on chameleons from `Multi-scale Attributed
    Node Embedding <https://arxiv.org/abs/1909.13021>`__, introduced by
    `Geom-GCN: Geometric Graph Convolutional Networks
```

**Review by frozenbugs — APPROVED** [review]

**Review by jermainewang — APPROVED** [review]

File: python/dgl/data/wiki_network.py (hunk 6 above)
Comment (mufeili) [reply python/dgl/data/wiki_network.py:113]: did "later processed by" -> "modified by"
